src/com/facebook/buck/jvm/java/SourceReadingPackageFinder.java

```diff
@@ -0,0 +1,278 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.jvm.java;
+
+import com.facebook.buck.io.MorePaths;
+import com.facebook.buck.io.ProjectFilesystem;
+import com.facebook.buck.jvm.core.JavaPackageFinder;
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.util.HumanReadableException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * Designed to parse just enough valid java to figure out the package name from a java source file.
+ */
+class SourceReadingPackageFinder implements JavaPackageFinder {
+
+  private final ProjectFilesystem filesystem;
+
+  public SourceReadingPackageFinder(ProjectFilesystem filesystem) {
+    this.filesystem = filesystem;
+  }
+
+  @Override
+  public Path findJavaPackageFolder(Path pathRelativeToProjectRoot) {
+    if ("java".equals(MorePaths.getFileExtension(pathRelativeToProjectRoot))) {
+      String packageName = findJavaPackage(pathRelativeToProjectRoot);
+
+      Iterator<String> segments = Splitter.on('.').split(packageName).iterator();
+      if (!segments.hasNext()) {
+        throw new IllegalStateException("Unable to create path from " + packageName);
+      }
+      Path path = pathRelativeToProjectRoot.getFileSystem().getPath(segments.next());
+      while (segments.hasNext()) {
+        path = path.resolve(segments.next());
+      }
+      return path;
+    }
+    throw new IllegalArgumentException("Only java source files are supported");
+  }
+
+  @Override
+  public String findJavaPackage(Path pathRelativeToProjectRoot) {
+    Path path = filesystem.resolve(pathRelativeToProjectRoot);
+
+    try (BufferedReader reader = Files.newBufferedReader(path)) {
+      return findJavaPackage(reader);
+    } catch (IOException e) {
+      throw new HumanReadableException(
+          "Unable to read package for file: %s",
+          pathRelativeToProjectRoot);
+    }
+  }
+
+  @Override
+  public String findJavaPackage(BuildTarget buildTarget) {
+    throw new UnsupportedOperationException("Unable to just guess a package: " + buildTarget);
+  }
+
+  @VisibleForTesting
+  String findJavaPackage(Reader reader) throws IOException {
+    Lexer lexer = new Lexer(reader);
+
+    for (;;) {
+      Token token = lexer.getNextToken();
+
+      switch (token.getKind()) {
+        case PACKAGE:
+          return token.getValue();
+
+        case UNKNOWN:
+          // We have no idea what to do. Assume the default package
+          return "";
```

Comment:
If we never see a package declaration, we're in the default package, and that might happen. We also use the `UNKNOWN` token to handle the case where we hit anything that might precede a package statement, so `interface`, `class`, and `enum` (and modifiers) all get treated that way. I'd chosen to play it safe for now, and the approach seems to work on both Buck's code and Selenium's